go_covert_lib/controller/channel/icmpIP/icmpIP.go

```diff
@@ -0,0 +1,155 @@
+package icmpIP
+
+import (
+	"../embedders"
+	"bytes"
+	"context"
+	"errors"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"golang.org/x/net/ipv4"
+	"net"
+	"sync"
+	"time"
+)
+
+// We make the fields public for logging
+type packet struct {
+	Ipv4h ipv4.Header
+	icmph  layers.ICMP
+}
+
+type syncPktMap struct {
+	mutex  *sync.Mutex
+	pktMap map[uint16][]packet
+}
+
+type acceptedConn struct {
+	conn net.Conn
+	friendPort uint16
+}
+
+type Config struct {
+	FriendIP          [4]byte
+	OriginIP          [4]byte
+	FriendReceivePort uint16
+	OriginReceivePort uint16
+	DialTimeout       time.Duration
+	Encoder           icmpEncoder
+
+	AcceptTimeout time.Duration
+
+	// The intra-packet read timeout. Set zero for no timeout.
+	// The receive method will block until a three way handshake
+	// is complete and the listener returns, and will only exit with a
+	// read timeout if the intra-packet delay is too large.
+	ReadTimeout time.Duration
+	// The timeout for writing the packet to a raw socket. Set zero for no timeout.
+	WriteTimeout time.Duration
+}
+
+type Channel struct {
+	conf    Config
+	rawConn *ipv4.RawConn
+	cancel  chan bool
+
+	// For debugging purposes, log all packets received and sent
+	sendPktLog    *syncPktMap
+	receivePktLog *syncPktMap
+
+	// We make the mutex a pointer to avoid the risk of copying
+	writeMutex *sync.Mutex
+	closeMutex *sync.Mutex
+
+	acceptChan chan acceptedConn
+}
+
+func (c *Channel) Close() error {
+	c.closeMutex.Lock()
+	defer c.closeMutex.Unlock()
+	select {
+	// Have we already closed
+	case <-c.cancel:
+		return nil
+	default:
+		close(c.cancel)
+	}
+	return c.rawConn.Close()
+}
+
+//create channel
+func MakeChannel(conf Config) (*Channel, error) {
+
+	c := &Channel{
+		conf:          conf,
+		cancel:        make(chan bool),
+		sendPktLog:    MakeSyncMap(),
+		receivePktLog: MakeSyncMap(),
+		writeMutex:    &sync.Mutex{},
+		closeMutex:    &sync.Mutex{},
+		// Only 32 connections can be accepted before they begin to be dropped
+		acceptChan: make(chan acceptedConn, maxAccept),
+	}
+
+	if c.conf.Encoder == nil {
+		c.conf.Encoder = &IDEncoder{}
+	}
+
+	//ip network within the ICMP protocol
+	conn, err := net.ListenPacket("ip4:icmp", "0.0.0.0")
+	if err != nil {
+		return nil, err
+	}
+
+	c.rawConn, err = ipv4.NewRawConn(conn)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	return c, nil
+}
+
```

Comment:
Have you run the tests for this? Have you run the controller tests? This doesn't seem to have a Receive method, which will cause the rest of the system to not compile.